namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerInterface.scala

```diff
@@ -333,34 +317,25 @@ class ThriftNamerInterface(
     }
   }
 
-  private[this] val addrCacheMu = new {}
-  private[this] var addrCache: Map[Path, AddrObserver] = Map.empty
-  private[this] def observeAddr(id: Path): AddrObserver =
-    addrCacheMu.synchronized {
-      addrCache.get(id) match {
-        case Some(obs) =>
-          Trace.recordBinary("namerd.srv/addr.cached", true)
-          obs
-
-        case None =>
-          Trace.recordBinary("namerd.srv/addr.cached", false)
-          val resolution = bindAddrId(id).run.flatMap {
-            case Activity.Pending => Var.value(Addr.Pending)
-            case Activity.Failed(e) => Var.value(Addr.Failed(e))
-            case Activity.Ok(tree) => tree match {
-              case NameTree.Leaf(bound) => bound.addr
-              case NameTree.Empty => Var.value(Addr.Bound())
-              case NameTree.Fail => Var.value(Addr.Failed("name tree failed"))
-              case NameTree.Neg => Var.value(Addr.Neg)
-              case NameTree.Alt(_) | NameTree.Union(_) =>
-                Var.value(Addr.Failed(s"${id.show} is not a concrete bound id"))
-            }
-          }
-          val obs = AddrObserver(resolution, stamper)
-          addrCache += (id -> obs)
-          obs
+  private[this] val addrCache = new ObserverCache[Path, Option[Addr.Bound]](
+    activeCapacity = 100,
+    inactiveCapacity = 110
+  )({ id =>
+    Trace.recordBinary("namerd.srv/addr.cached", false)
+    val resolution = bindAddrId(id).run.flatMap {
+      case Activity.Pending => Var.value(Addr.Pending)
+      case Activity.Failed(e) => Var.value(Addr.Failed(e))
+      case Activity.Ok(tree) => tree match {
+        case NameTree.Leaf(bound) => bound.addr
+        case NameTree.Empty => Var.value(Addr.Bound())
+        case NameTree.Fail => Var.value(Addr.Failed("name tree failed"))
+        case NameTree.Neg => Var.value(Addr.Neg)
+        case NameTree.Alt(_) | NameTree.Union(_) =>
+          Var.value(Addr.Failed(s"${id.show} is not a concrete bound id"))
       }
     }
+    AddrObserver(resolution, stamper)
+  })
```

Comment:
similarly, i think this would be a bit clearer as a separate function rather than inlined.